### DISCUSSION TITLE
cli: add config option to disable automatic uploading of asciicasts

### DISCUSSION THREAD

**Discussion by cbarcenas** [asciinema/discussions#21] [Ideas]
Currently, it's too easy for users to accidentally publish sensitive screencasts to asciinema.org: all you need to do is accidentally omit a filename argument and then press <Enter> at the upload confirmation prompt.

At many security-minded organizations, particularly those that regularly deal with sensitive information, the automatic upload feature of `asciinema rec` would turn asciinema into a data leakage risk. This may in turn prevent its use at such organizations.

I propose that a configuration option, `upload`, be added to the `[record]` section of the config file. This option, which would default to `yes`, could be used to control the automatic upload functionality of `asciinema rec`. If set to `no`, automatic upload functionality would be disabled and the filename argument to `asciinema rec` would become mandatory.

I have implemented this proposal and will submit a PR for review.

**Comment by ku1ik** [comment]
As this is linked to asciinema/asciinema#178 I'm closing it too.

**Comment by saxophonicle** [comment]
this issue was never fixed it needs to be reopened. The PR it was linked to was closed without merging and so this issue is lost to the void.

**Comment by AkihiroSuda** [comment]
@sickill Could you reopen this issue as @saxophonicle said?

**Comment by ku1ik** [comment]
My comments in the referenced PR were never addressed, and I never heard back from cbarcenas. Thus both PR and issue were closed.

**Comment by ku1ik** [comment]
If you'd like to completely prevent uploads today you can put this in the config file, as a temporary workaround:

```ini
[api]
url = http://nope
```

**Reply by ku1ik** [reply]
Or set `ASCIINEMA_API_URL=http://nope` variable (e.g. in your shell config).

**Comment by ghost** [comment]
We're having the same issue as @cbarcenas: We deal with sensitive data on a daily basis and the risk of automatically uploading screen casts is so significant that we are currently refraining from using asciinema, even though it's a really useful tool.

Therefore I'd offer to implement a solution. I would suggest re-implementing asciinema/asciinema#178 (re)based on the current state of the repository, i.e. a configuration option that can be used to disable uploading. Additionally, I would also have asciinema look for the configuration at `/etc/asciinema/config` such that out administrators can roll out a centralized solution.

@sickill Does that sound okay to you?

**Comment by IAXES** [comment]
> 
> 
> this issue was never fixed it needs to be reopened. The PR it was linked to was closed without merging and so this issue is lost to the void.

I frequently use this tool for tutorials, examples, etc.; and it's pretty awesome. That being said, I too was very concerned about accidental data leakage (i.e. there are use cases where this tool couldn't be used for legal reasons, solely due to the **potential** for accidental data leakage by inexperienced users) or a simple [mis-configuration problem](https://owasp.org/www-project-top-ten/2017/A6_2017-Security_Misconfiguration) resulting in sensitive data being leaked (i.e. default behaviour is to allow uploads, so mis-configuration results in potential for uploads in the worst case, or the app just not launching in the best case scenario).

In my own use case, the solution was to put all of the tooling needed for `asciinema` (plus the app itself) into a Docker container, and providing a launcher script (i.e. shell script and/or `docker-compose.yaml`) for the container so that it was always run with no internet access (there's a couple ways to do this; I just took the heavy-handed approach and disabled networking entirely; the container also has some "sanitized" launcher scripts inside so as to limit the number of ways in which `asciinema` can be launched, plus the script used in the `Dockerfile` `ENTRYPOINT` does a quick check via `iproute2`/`ip` for network interfaces, and "deliberately dies early" if anything other than loopback is detected, as a safety feature in case someone launches the container directly rather than via the launcher scripts I provided). This was enough for my own requirements.

This might help others aiming to lock down output options for the tool without going so far as to patch + repackage/re-deploy it.

Cheers!

**Reply by JaneX8** [reply]
Great answer.

**Comment by bkil** [comment]
Duplicated by:
- asciinema/asciinema#263
- asciinema/asciinema#316

**Comment by LudwikJaniuk** [comment]
> If you'd like to completely prevent uploads today you can put this in the config file, as a temporary workaround:
> 
> ```ini
> [api]
> url = http://nope
> ```

This totally works for me :) Perhaps this could just be added to the docs, to make it easier to find for people?

**Comment by ku1ik** [comment]
Proper fix, https://github.com/asciinema/asciinema/pull/576, is coming in v2.4.0, which is around the corner.

**Reply by JaneX8** [reply]
This is not a proper fix for the issue in my opinion. Or I fail to understand it. I explain here: https://github.com/orgs/asciinema/discussions/21#discussioncomment-16058041.

**Comment by ku1ik** [comment]
Just released 2.4.0 with improved prompt which requires explicit choice: https://github.com/asciinema/asciinema/releases/tag/v2.4.0

**Comment by ku1ik** [comment]
I have reworked the upload related behavior for the upcoming asciinema 3.0:

1. `asciinema rec` now always requires a filename and never does any uploading - to upload you will use explicit `asciinema upload <filename>`
2. There's no implicit asciinema server URL configured - on first use (first `upload` or `auth` command) you will see a prompt asking for a server URL, with `asciinema.org` being suggested. Once confirmed the server URL will be saved and reused for future uploads.

I hope this addresses most of the concerns. You won't be able to accidentally upload anything anywhere, because it will require intent (`asciinema upload`), _and_ a configuration step (server URL prompt).

When presented with a prompt you would confirm the use of asciinema.org or you would enter the URL of your self-hosted asciinema server instance. You will still be able to pre-configure the server URL in the config file via `server.url = "..."` to avoid the prompt.

3.0 hasn't been released yet, but it's close. If you'd like to test it then here's a release candidate (also providing binary downloads for Linux and macOS): https://github.com/asciinema/asciinema/releases/tag/v3.0.0-rc.1

Let me know if this looks like a good solution. Thanks!

**Reply by dezza** [reply]
but now an auto upload command is lacking, correct? Missing that.

**Reply by JaneX8** [reply]
@ku1ik This is much better, however I suggest still one security improvement. Do not upload at all if auth has never been used before. If `asciinema upload` is explicitly triggered, it should fail and suggest to use `asciinema auth` first. Moreover, there should be an `deauth` or `reauth` option perhaps?

@dezza An autoupload function is a very bad idea for all the mentioned security reasons. However if such feature is ever provided, above should REALLY apply.

**Reply by dezza** [reply]
I don't disagree on that part, anonymous auto uploads were an unexpected feature.

However, since the service is already more or less tightly coupled to a web app- I don't see a reason not to provide --autoupload or whatever.

Granted, I did just create a 2 command pipeline as alias, but handling saving file and uploading just seems.. Unintegrated.. Inaccessible.

**Comment by LudwikJaniuk** [comment]
@ku1ik That sounds good to me!

**Comment by LudwikJaniuk** [comment]
And thanks for doing this.

**Comment by ku1ik** [comment]
What I described in my previous post has been released in asciinema CLI 3.0, therefore I believe we can close this topic.

**Comment by JaneX8** [comment]
I am very new to asciinema and looking into disabling (accidental) upload too. My background is cyber security. As https://github.com/orgs/asciinema/discussions/21#discussioncomment-5717564 mentioned too, there is a potential for accidental data leakage and as cast files clearly can contain very sensitive data, this is a security concern. For me personally it's a reason not to recommend this awesome tool, just yet.

https://github.com/asciinema/discussions/issues/20#issuecomment-1521486172 is not a good solution in my opinion as it will (i think) try to resolve the domain and potentially pollute some logs. It seems there is no good solution still, except for setting a non-routable domain as an upload URL.

I am suggesting something broader. I am not familiar with the asciicast format but I would suggest adding a parameter to the recorder, like `--dontupload` (or another naming but it should be clear from the naming it's a soft control, not enforceable) flag on `asciinema upload` that would embed a mark in the asciiformat metadata. I thought about `--preventupload` but it suggest more of a hard control, which is unenforceable. Then everytime the `asciinema upload` is used, it would simply say that that file was recorded explicitly with the flag and is therefore not **recommended** to be uploaded, hence upload using asciinema is disabled. It should then be documented however that the file of course can be modified, manually uploaded or shared anyway. Hence my mark on it being a soft control. I think this is somewhat inspired by something NPM has in package.json files.

If however a non-routable domain is still the solution, I recommend using a domain with a top level domain (TLD) for exactly that purpose as described in [RFC 2606 section 2](https://datatracker.ietf.org/doc/html/rfc2606#section-2) using the `.invalid` TLD.

```bash
# Temporary disable upload for one terminal session
# https://docs.asciinema.org/manual/server/self-hosting/#__tabbed_1_1
export ASCIINEMA_SERVER_URL=http://asciinema.invalid # CLI 3.x
export ASCIINEMA_API_URL=http://asciinema.invalid # CLI 2.x
# Make it permanent by adding these two lines to ~/.bashrc or ~/.bash_profile or ~/.zshrc.
# Security fallback (in case of misconfig) add the domain asciinema.org to /etc/hosts or managed DNS and route it to 127.0.0.1

# Test uploading
asciinema rec test
exit
asciinema upload test.cast
# asciinema: upload failed: <urlopen error [Errno -2] Name or service not known>
# asciinema: retry later by running: asciinema upload test.cast
```

_I am not sure what the default upload URL is but I think https://asciinema.org, so an option is also to add that domain to `/etc/hosts` and make it non-routable by pointing to 127.0.0.1 for example, but it's for the same and other reasons not a good idea._
